rusk/src/lib/lib.rs

```diff
@@ -543,6 +511,135 @@ fn accept(
+///
+/// 1. Call the "spend" function on the transfer contract with unlimited gas. If
+///    this fails, the transaction should be considered invalid, or unspendable,
+///    and an error is returned.
+///
+/// 2. If the transaction includes a contract call, execute it with the gas
+///    limit given in the transaction, or with the block gas remaining,
+///    whichever is smallest. If this fails with an out of gas, two possible
+///    things happen:
+///        * We use the transaction gas limit and will treat this as any other
+///          transaction.
+///        * We used the block gas remaining and can't be sure of what to do. In
+///          this case we return early with an [TxError::BlockLimitAfter], since
+///          we are in a bad state, and can't be sure of what to do.
+///    For any other transaction error we proceed to step 3.
+///
+/// 3. Call the "refund" function on the transfer contract with unlimited gas.
+///    The amount charged depends on if the transaction has executed a call or
+///    not. If it has there are two cases:
+///        * The call succeeded and the transaction will be charged for gas used
+///          plus the amount charged by a transaction of its type.
+///        * The call errored and the transaction will be charged the full gas
+///          given.
+///    If the transaction has not executed a call only be the amount charged for
+///    a transaction of its type.
+fn execute(
+    session: &mut Session,
+    tx: &Transaction,
+    block_gas_left: u64,
+) -> Result<(Vec<u8>, u64), TxError> {
+    let gas_for_spend = spent_gas_per_input(tx.nullifiers.len());
+
+    // If the gas given is less than the amount the node charges per input, then
+    // the transaction is unspendable.
+    if tx.fee.gas_limit < gas_for_spend {
+        return Err(TxError::Unspendable(PiecrustError::OutOfPoints));
+    }
+
+    // If the gas to spend is more than the amount remaining in a block, then
+    // the transaction can't be spent at this spot in the block.
+    if block_gas_left < gas_for_spend {
+        return Err(TxError::BlockLimit(PiecrustError::OutOfPoints));
+    }
+
+    // Spend the transaction. If this error the transaction is unspendable.
+    session.set_point_limit(u64::MAX);
+    session
+        .call(TRANSFER_CONTRACT, "spend", tx)
+        .map_err(TxError::Unspendable)?;
+
+    let mut gas_spent = gas_for_spend;
+
+    let block_gas_left = block_gas_left - gas_spent;
+    let tx_gas_left = block_gas_left - gas_spent;
+
+    let res = tx
+        .call
+        .as_ref()
+        .map(|(contract_id_bytes, fn_name, fn_data)| {
+            let contract_id = ContractId::from_bytes(*contract_id_bytes);
+            let gas_remaining = cmp::min(block_gas_left, tx_gas_left);
+
+            session.set_point_limit(gas_remaining);
+
+            match session.call_raw(contract_id, fn_name, fn_data.clone()) {
+                Ok(vec) => {
+                    gas_spent += session.spent();
```

Comment:
Is this `session.spent()` including the gas spent while `spend` call?

If so, we should subtract the real `spend`'s gas_spent

Reply:
Fortunately no, it doesn't. It's the gas spent by the last call.